substrate/frame/meta-tx/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+//! [`frame_system::CheckGenesis`], [`frame_system::CheckMortality`], [`frame_system::CheckNonce`],
+//! etc., are applicable in the context of meta transactions.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub mod weights;
+#[cfg(feature = "runtime-benchmarks")]
+pub use benchmarking::types::{BenchmarkHelper, BenchmarkHelperFor, WeightlessExtension};
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+use core::ops::Add;
+use frame_support::{
+	dispatch::{DispatchInfo, GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+};
+use frame_system::{pallet_prelude::*, RawOrigin as SystemOrigin};
+use sp_runtime::traits::{
+	AsTransactionAuthorizedOrigin, DispatchTransaction, Dispatchable, IdentifyAccount,
+	TransactionExtension, Verify,
+};
+use sp_std::prelude::*;
+
+/// Meta Transaction type.
+///
+/// The data that is provided and signed by the signer and shared with the relayer.
+#[derive(Encode, Decode, PartialEq, Eq, TypeInfo, Clone, RuntimeDebug)]
+pub struct MetaTx<Address, Signature, Call, Extension> {
+	/// The signer's address.
+	address: Address,
+	/// The signature of the meta transaction.
+	signature: Signature,
+	/// The target call to be executed on behalf of the signer.
+	call: Call,
+	/// The extension/s for the meta transaction.
+	extension: Extension,
+}
+
+impl<Address, Signature, Call, Extension> MetaTx<Address, Signature, Call, Extension> {
+	/// Create a new meta transaction.
+	pub fn new(address: Address, signature: Signature, call: Call, extension: Extension) -> Self {
+		Self { address, signature, call, extension }
+	}
+}
+
+/// The [`MetaTx`] for the given config.
+pub type MetaTxFor<T> = MetaTx<
+	<<T as Config>::PublicKey as IdentifyAccount>::AccountId,
+	<T as Config>::Signature,
+	<T as Config>::RuntimeCall,
+	<T as Config>::Extension,
+>;
+
+/// The payload that has been signed for the [`MetaTx`].
+pub type SignedPayloadFor<T> =
+	sp_runtime::generic::SignedPayload<<T as Config>::RuntimeCall, <T as Config>::Extension>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Weight information for calls in this pallet.
+		type WeightInfo: WeightInfo;
+		/// The overarching origin type.
+		// We need extra `AsTransactionAuthorizedOrigin` bound to use `DispatchTransaction` impl.
+		type RuntimeOrigin: AsTransactionAuthorizedOrigin
+			+ From<SystemOrigin<Self::AccountId>>
+			+ IsType<<Self as frame_system::Config>::RuntimeOrigin>;
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The overarching call type.
+		type RuntimeCall: Parameter
+			+ GetDispatchInfo
+			+ Dispatchable<
+				Info = DispatchInfo,
+				PostInfo = PostDispatchInfo,
+				RuntimeOrigin = <Self as Config>::RuntimeOrigin,
+			> + IsType<<Self as frame_system::Config>::RuntimeCall>;
```

Comment:
I think we no longer need to copy types like this. We can get rid of `RuntimeCall` and `RuntimeOrigin` associated types here and directly bound the frame system config supertrait:

```rust
frame_system::Config<
    RuntimeCall: Parameter
			+ GetDispatchInfo
			+ Dispatchable<
				Info = DispatchInfo,
				PostInfo = PostDispatchInfo,
				RuntimeOrigin = <Self as Config>::RuntimeOrigin,
			>,
    RuntimeOrigin: AsTransactionAuthorizedOrigin
			+ From<SystemOrigin<Self::AccountId>>
>
```

Reply:
these types of bounds are unstable, CI jobs failing, I left them unchanged for now

Reply:
It should be stable since June :-/ https://blog.rust-lang.org/2024/06/13/Rust-1.79.0.html

Reply:
Was there a particular CI jobs failing? I will try to fix it.

For now it seems somewhat green: https://github.com/paritytech/polkadot-sdk/pull/6817